### PR TITLE
Add a possibility to retrieve current location from the device

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
Addresses https://github.com/appium/appium/issues/5047

@vmaxim Can you please take a look?

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -10,7 +10,6 @@
     <uses-permission android:name="android.permission.CHANGE_WIFI_STATE" />
     <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
-    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply app/src/main/AndroidManifest.xml]: I assume coarse location permission sis granted automatically if FINE permission is given. But ok, I'll remove it from there

File: app/src/main/java/io/appium/settings/utils/GPSTracker.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        this.mContext = context;
+    }
+
+    public Location getLocation() {
+        try {
+            final LocationManager locationManager =
+                    (LocationManager) mContext.getSystemService(LOCATION_SERVICE);
+            boolean isGPSEnabled = false;
+            try {
+                isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
```
Comment (vmaxim) [inline app/src/main/java/io/appium/settings/utils/GPSTracker.java]: I'd prefer to use [PASSIVE_PROVIDER](https://developer.android.com/reference/android/location/LocationManager.html#PASSIVE_PROVIDER)

File: app/src/main/java/io/appium/settings/utils/GPSTracker.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            }
+
+            if (!isGPSEnabled && !isNetworkEnabled) {
+                Log.i(TAG, "Both GPS and Network providers are disabled");
+                return null;
+            }
+
+            Location location = null;
+            if (isGPSEnabled) {
+                locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER,
```
Comment (vmaxim) [inline app/src/main/java/io/appium/settings/utils/GPSTracker.java]: I assume that this implementation is rather unstable - some location updates can be missed.
To catch all updates `requestLocationUpdates` should be called in `onCreate()` method. And the service should handle updates in `onLocationChanged` method.
Like a:
```java
public class GPSTracker extends Service implements LocationListener {
    private static final String TAG = GPSTracker.class.getSimpleName();

    private static Location location = null;
    LocationManager locationManager;
    
    @Override
    public void onCreate() {
        addListenerLocation();
    }
    
    @Override
    public void onDestroy() {
        super.onDestroy();
        locationManager.removeUpdates(this);
    }

    private void addListenerLocation() {
        locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);
        locationManager.requestLocationUpdates(LocationManager.PASSIVE_PROVIDER,
                0, 0, this);
    }
    
    public static Location getLocation() {
        return location;
    }

    @Override
    public void onLocationChanged(Location location) {
        this.location = location;
    }
}
```

Comment (mykola-mokhnach) [reply app/src/main/java/io/appium/settings/utils/GPSTracker.java]: is it mandatory to register the tracker class in the manifest if we change it this way?

Comment (vmaxim) [reply app/src/main/java/io/appium/settings/utils/GPSTracker.java]: Yes, it should be a service like a ```LocationService```.

Comment (mykola-mokhnach) [reply app/src/main/java/io/appium/settings/utils/GPSTracker.java]: changed

File: app/src/main/java/io/appium/settings/GPSTracker.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Override
+    public void onCreate() {
+        addLocationListener();
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (locationManager != null) {
```
Comment (vmaxim) [inline app/src/main/java/io/appium/settings/GPSTracker.java]: location = null;

**Comment by mykola-mokhnach** [comment]
I cannot make it to return proper coordinates on a device

**Comment by vmaxim** [comment]
Works fine for me on real device 5.0.2

**Comment by mykola-mokhnach** [comment]
I tried Android 8

**Comment by mykola-mokhnach** [comment]
Install, grant permissions, execute app, execute broadcast command

**Comment by vmaxim** [comment]
Have you launched GPSTracker service?

**Pull request closed without merging**